### PR TITLE
Add support for the Appsody-Operator

### PR DESCRIPTION
**Pull request by seabaylea**
This adds support for the Appsody-operator into the deploy commands by:
* Adding a new `operator` command with `install` and `uninstall` to support installing and uninstalling the operator
* Updating the `deploy` command to work with a default `app-deploy.yaml` that is extracted from inside the Appsody Stack associated with the project
  * This falls back to the existing Knative Serving approach if no file is found in the stack
* Adding a `deploy --generate` flag to generate the `app-deploy.yaml` but not deploy the app
* Add a new `delete` command to remove the deployed project

### CODE REVIEW

File: cmd/deploy.go [hunk 1]
```diff
@@ -15,107 +15,395 @@
...
 
-var namespace, tag string
-var push bool
+var configFile, namespace, operatorspace, watchspace, tag string
+var generate, force, push bool
 
 var deployCmd = &cobra.Command{
 	Use:   "deploy",
-	Short: "Build and deploy your Appsody project on a local Kubernetes cluster",
+	Short: "Build and deploy your Appsody project to Kubernetes cluster",
```
Comment (kewegner) [inline cmd/deploy.go]: to "a" or "your" Kubernetes cluster???

File: cmd/deploy.go [hunk 2]
```diff
@@ -15,107 +15,395 @@
...
-				if err != nil {
-					Warning.log("The exposed port is not a valid integer. The service will not be accessible.")
-					port = 0
+		}
+
+		exists, err := exists(configFile)
+		if err != nil {
+			return errors.Errorf("Error checking status of %s", configFile)
+		}
+		if !exists || (exists && force == true) {
```
Comment (kewegner) [inline cmd/deploy.go]: force== true triggers a lint error

File: cmd/deploy.go [hunk 3]
```diff
@@ -15,107 +15,395 @@
...
+			if os.IsNotExist(err) {
+				return errors.Errorf("Config file does not exist %s. ", configFile)
+			}
+			return errors.Errorf("Failed reading file %s", configFile)
+		}
+		defer yamlFile.Close()
+		scanner := bufio.NewScanner(yamlFile)
+		scanner.Split(bufio.ScanLines)
+		var txtlines []string
+		for scanner.Scan() {
```
Comment (kewegner) [inline cmd/deploy.go]: I assume the yaml structure is too complicated to create structs to parse it like we do the index.yaml files?

File: cmd/deploy.go [hunk 4]
```diff
@@ -15,107 +15,395 @@
...
+		if portsErr != nil {
+			return portsErr
+		}
+		if len(portsStr) == 0 {
+			//No ports exposed
+			Warning.log("This container exposes no ports. The service will not be accessible.")
+			port = 0 //setting this to 0
+		} else {
+			portStr := portsStr[0]
+			Warning.log("Picking the first exposed port as the KNative service port. This may not be the correct port.")
```
Comment (kewegner) [inline cmd/deploy.go]: I agree with your comment.

File: cmd/deploy.go [hunk 5]
```diff
@@ -15,107 +15,395 @@
...
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		err := dockerStop(containerName)
+		if err != nil {
+			Error.log(err)
+		}
+		//dockerRemove(containerName) is not needed due to --rm flag
```
Comment (kewegner) [inline cmd/deploy.go]: I do not see any --rm flag, is it required?  
There is no wait after dockerStop, is one needed (similar to appsody run?)

File: cmd/deploy.go [hunk 6]
```diff
@@ -15,107 +15,395 @@
...
+		os.Exit(1)
+	}()
+	cmdName = "docker"
```
Comment (kewegner) [inline cmd/deploy.go]: Future: no change needed, this code looks like the code we have in extract, I wonder if there is a potential here for future refactoring?

File: cmd/utils.go [hunk 7]
```diff
@@ -542,38 +542,21 @@ func GenKnativeYaml(yamlTemplate string, deployPort int, serviceName string, dep
 		return "", err
```
Comment (kewegner) [inline cmd/utils.go]: FUTURE: I think in future refactoring we might want to consider moving the k8s functions out of utils.go into a k8utils.go etc, like we are planning on doing for docker commands.

File: cmd/operator.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+	"github.com/pkg/errors"
+
+	"github.com/spf13/cobra"
+)
+
+var operatorYamlName = "appsody-app-operator.yaml"
+var appsodyCRDName = "appsody-app-crd.yaml"
+var crdURL = "https://raw.githubusercontent.com/appsody/appsody-operator/master/deploy/releases/0.1.0/appsody-app-crd.yaml"
+var operatorURL = "https://raw.githubusercontent.com/appsody/appsody-operator/master/deploy/releases/0.1.0/appsody-app-operator.yaml"
```
Comment (chilanti) [inline cmd/operator.go]: Are we going with these URLs? I thought you said we'd point to the latest release.

File: cmd/deploy.go [hunk 9]
```diff
@@ -15,107 +15,395 @@
...
+		return errors.Errorf("Failed to write local application configuration file: %s", err)
+	}
+
+	Info.log("Created deployent manifest: ", configFile)
+	return nil
+}
+
 func init() {
 	rootCmd.AddCommand(deployCmd)
+	deployCmd.PersistentFlags().BoolVar(&generate, "generate", false, "Only generate the deployment configuration file. Do not deploy the project.")
```
Comment (chilanti) [inline cmd/deploy.go]: Would `--generate-only` or `--generate-manifest` or `--generate-yaml` be more meaningful?

File: cmd/deploy.go (hunk 1 above)
Comment (seabaylea) [reply cmd/deploy.go]: Fixed.

File: cmd/deploy.go (hunk 2 above)
Comment (seabaylea) [reply cmd/deploy.go]: Fixed.

File: cmd/deploy.go (hunk 3 above)
Comment (seabaylea) [reply cmd/deploy.go]: The challenge is that the CRD spec for the app-deploy.yaml is currently v1alpha1 so it likely to change (whereas the use of the specific `applicationImage:` field is unlikely to).

Ultimately we'd like to import the YAML definition as a struct from the appsody-operator project and then parse it using that directly.

File: cmd/deploy.go (hunk 5 above)
Comment (seabaylea) [reply cmd/deploy.go]: This is actually a copy/paste from the original deploy command - so presumably its correct ;-P

File: cmd/operator.go (hunk 8 above)
Comment (seabaylea) [reply cmd/operator.go]: These are currently correct and work - yes, we will need to switch this to the formal Git release when its present (we should ideally get that PR in before we cut a new release of the CLI).

File: cmd/deploy.go (hunk 9 above)
Comment (seabaylea) [reply cmd/deploy.go]: Switched to use `--generate-only`

**Comment by seabaylea** [comment]
@chilanti @kewegner Thanks for the review. I think I've addressed everything raised.

Can you take another look?

**Comment by chilanti** [comment]
After talking to Chris - there is enough `dry-run` support (it's buried in the lower level functions such as `KubeGet` as it should be).

**Comment by kewegner** [comment]
Merging we need to reenable the requirement for tests to pass before merging
so we can merge i

**Pull request merged**